### ISSUE TITLE
changing between clients is too slow

### ISSUE THREAD

**Issue by DeathRowTools** [valyala/fasthttp#1223]
guess what ! im back again with another question :)
so before i asked you about the fastest way to run my program on high requests per second 
and you gave me this code:
`clientsMu.Lock()
		client, ok := clients[proxy]
		if !ok {
			client = &fasthttp.Client{
				Dial: FasthttpHTTPDialer(proxy),
				ReadBufferSize: 8192,
			}
			clients[proxy] = client
		}
		clientsMu.Unlock()`

the code is working great and it gives me 40000 request per second 
which is by far the best libary i have ever used "for sending requests"
but my problem here is that when i run 40k r/s my proxies gets blocked often and i need to clear the clients every 15 seconds to conntect to an fresh ip "im using rotating proxies"
here is my code: 

`func clearclients() {
	for true {
		time.Sleep(15 * time.Second)
		clients = make(map[string]*fasthttp.Client)

	}
}`

this will clear all old clients and create new ones

but the problem here that it takes time each time i make an client clear
my r/s goes from 40k to 4k and stays at 4k for 5 seconds then it up to 20k then 40k
then again and again after 15 seconds same thing happen
my script needs to stay stable at 40k r/s without going under that for more than 1 second
so i tried an solution 
to make 2 clients maps and i fill the first one with clients 
and use it for 15  second
and meanwhile the second clients will fill clients into it
then it will changes to clients number 2 map
so what should happen "or what i thought at least"
that if i pre connected to proxies with my clients number 2
it should be up and ready for sending 40k r/s same second 
but the problem here is that i still face the same problem each time i change from client 1 map to client 2 map the r/s goes down to 4k 
here is my code:

`func clearclients() {
		for true {
			time.Sleep(time.Second * 20)
			var clients2 = make(map[string]*fasthttp.Client)
			for _,proxy := range proxies{
				clientsMu.Lock()
				client, ok := clients2[proxy]
				if !ok {
					client = &fasthttp.Client{
						Dial: FasthttpHTTPDialer(proxy),
						ReadBufferSize: 8192,
					}
					clients2[proxy] = client
				}
				clientsMu.Unlock()
			}
			clients = clients2
			time.Sleep(10)

			var clients1 = make(map[string]*fasthttp.Client)
			for _,proxy := range proxies{
				clientsMu.Lock()
				client, ok := clients1[proxy]
				if !ok {
					client = &fasthttp.Client{
						Dial: FasthttpHTTPDialer(proxy),
						ReadBufferSize: 8192,
					}
					clients1[proxy] = client
				}
				clientsMu.Unlock()
			}
			clients = clients1
		}
	}`
	
with this func it should every 15 seconds change between clients and its working but its slow 
i thought that what makes the script slower is that each time i clear my clients the script needs to re make clients and fill it into the map
but if i already make that before sending requests with it why is it too slow ? 
does the  client maps like save the connection between request and just re use it? thats why its slow even when i fill clients into client map without using requests? it only goes 40k r/s after i make an few requests to my target 
thank you for any help in advance ! much appropriated <3

**Comment by erikdubbelboer** [comment]
Creating and start using the new client will always slow things down. The new client will have to setup new connections and do TLS handshakes which is slow. And with that amount of requests per second it probably has multiple connections per client.

What I think would be best is to start using the new client slowly over time instead of switching to it at once. So something like this (completely untested code):
```go
	var clientsMu sync.Mutex
	clients := make(map[string]*fasthttp.Client)
	newClients := make(map[string]*fasthttp.Client)
	newClientCreatedAt := make(map[string]time.Time)

	go func() {
		for {
			time.Sleep(time.Second)

			clientsMu.Lock()
			for proxy, at := range newClientCreatedAt {
				// After 100 seconds we are completely using the new client, so swap it to the clients map so newClients can start slowly with a new client again.
				if time.Now().Sub(newClientCreatedAt).Seconds() >= 100 {
					if client, ok := clients[proxy]; ok {
						client.CloseIdleConnections()
					}
					clients[proxy] = newClients[proxy]
					delete(newClients, proxy)
					delete(newClientCreatedAt, proxy)
				}
			}
			clientsMu.Unlock()
		}
	}()

	clientsMu.Lock()
	client, ok := clients[proxy]
	if !ok {
		client = &fasthttp.Client{
			Dial: FasthttpHTTPDialer(proxy),
			ReadBufferSize: 8192,
		}
		clients[proxy] = client
	} else {
		d := time.Now().Sub(newClientCreatedAt).Seconds()
		if d > 100 {
			d = 100
		}

		// Randomly start using the new proxy over a period of 100 seconds.
		if rand.Intn(100) < int(d) {
			client, ok = newClients[proxy]
			if !ok {
				client = &fasthttp.Client{
					Dial: FasthttpHTTPDialer(proxy),
					ReadBufferSize: 8192,
				}
				newClients[proxy] = client
				newClientCreatedAt[proxy] = time.Now()
			}
		}
	}
	clientsMu.Unlock()
```

**Comment by DeathRowTools** [comment]
Thanks for your efforts erik
But how can i make this code to replace all clients with new ones in less
than 15 seconds?
Im using 5000 proxies
And yes each client sends around 8 request per second
so with this method it will take 100 seconds to change the old clients to
new one by this time my old clients gonna gives me a lot of spam blocks.
So how can i make this same code but faster , in 15-20 second max.

**Comment by erikdubbelboer** [comment]
I would change the 100 to 7 in my example. Then each 15 seconds all requests will switch to a new client. 7 seconds from `clients` and another 7 seconds from `newClients`.

**Comment by DeathRowTools** [comment]
and i also noticed that when i dont change clients , use the same clients without clearing old ones
"for an different script" that dosent spam block alot.
after i hit 400k attempt using same clients
i start getting errors
is it due keep alive time out?
or how can i fix this also?

**Comment by erikdubbelboer** [comment]
It needs to be `newClientCreatedAt[proxy]`.

Are you sure it's an empty response with a 200 status? Sounds like an issue on the server, maybe it can't handle that many requests per second? Keep-alive timeouts or any other timeouts wouldn't cause this.

**Comment by DeathRowTools** [comment]
we are talking about instagram servers im sure they can handle millions of requests :)

**Comment by erikdubbelboer** [comment]
Yes but I'm also sure they have some form of spam protection when a lot of requests are coming from the same IP. They are too big a target not to have something like that.

What if you lower your requests per second by lot, do the empty responses still happen then after a while?

**Comment by DeathRowTools** [comment]
yes it exactly happens after 2:43 minutes
doesnt matter if im sending 50k r/s or 10 r/s 
always happends at the same time
thats why i thought its sort of time out for keep alive or something

**Comment by erikdubbelboer** [comment]
What if you send 10 requests per second? Some reasonable amount that a normal client could generate. Does it still happen?

There shouldn't be any timeout that could cause that to happen. It's not something in your own code happening at that point?

**Comment by DeathRowTools** [comment]
yes even if 10 request per second or even 1 it always happens

**Comment by DeathRowTools** [comment]
im struggling to use your function
![image](https://user-images.githubusercontent.com/91473167/155828006-874d4807-dca5-467c-90d0-4c427dfa1f21.png)

**Comment by erikdubbelboer** [comment]
If it even happens at 10 requests per seconds than something else is wrong. Would you be willing to share your code with me so I can have a look at what might be going wrong?

I can't see enough from that screenshot.

**Comment by DeathRowTools** [comment]
im using private api's i cant share them 
if you can give me your email or your instagram i can send them there?

**Comment by erikdubbelboer** [comment]
You can send it to erik@dubbelboer.com
